controllers/imagerepository_controller.go

```diff
@@ -72,55 +73,85 @@ func (r *ImageRepositoryReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
 
 	ref, err := name.ParseReference(imageRepo.Spec.Image)
 	if err != nil {
-		imageRepo.Status.LastError = err.Error()
-		if err := r.Status().Update(ctx, &imageRepo); err != nil {
-			return ctrl.Result{}, err
+		status := imagev1alpha1.SetImageRepositoryReadiness(
+			imageRepo,
+			corev1.ConditionFalse,
+			imagev1alpha1.ReconciliationFailedReason,
+			err.Error(),
+		)
+		if err := r.Status().Update(ctx, &status); err != nil {
+			return ctrl.Result{Requeue: true}, err
 		}
 		log.Error(err, "Unable to parse image name", "imageName", imageRepo.Spec.Image)
-		return ctrl.Result{}, nil
+		return ctrl.Result{}, err
 	}
 
-	canonicalName := ref.Context().String()
-	imageRepo.Status.CanonicalImageName = canonicalName
+	imageRepo.Status.CanonicalImageName = ref.Context().String()
 
 	now := time.Now()
-	ok, when := r.shouldScan(&imageRepo, now)
+	ok, when := r.shouldScan(imageRepo, now)
 	if ok {
 		ctx, cancel := context.WithTimeout(ctx, scanTimeout)
 		defer cancel()
-		tags, err := remote.ListWithContext(ctx, ref.Context()) // TODO: auth
-		if err != nil {
-			imageRepo.Status.LastError = err.Error()
-			if err = r.Status().Update(ctx, &imageRepo); err != nil {
-				return ctrl.Result{}, err
-			}
+
+		reconciledRepo, reconcileErr := r.scan(ctx, imageRepo, ref)
+		if err = r.Status().Update(ctx, &reconciledRepo); err != nil {
+			return ctrl.Result{Requeue: true}, err
 		}
 
-		imageRepo.Status.LastScanTime = &metav1.Time{Time: now}
-		imageRepo.Status.LastScanResult.TagCount = len(tags)
-		imageRepo.Status.LastError = ""
-		// share the information in the database
-		r.Database.SetTags(canonicalName, tags)
-		log.Info("successful scan", "tag count", len(tags))
-		if err = r.Status().Update(ctx, &imageRepo); err != nil {
-			return ctrl.Result{}, err
+		if reconcileErr != nil {
+			return ctrl.Result{RequeueAfter: when}, reconcileErr
```

Comment:
I'm not clear on whether controller-runtime will ignore the RequeueAfter (because there's an error) or respect it. If it's the latter, a comment on why it's being requeued at the scheduled time, rather than letting it back off, might make sense.

Reply:
I think we could return an error type from scan to signal rate limit errors and requeue right away, so that the runtime backoff with kick in. For other types of errors we could use the interval like we do for sources/helmreleases/kustomizations/etc.

Reply:
Yeah that's not a bad idea, to let the backoff kick in, if we get rate limited.
In the context of this PR -- is it useful to return both an error and `ctrl.Result{RequeueAfter: when}`?

Reply:
I've changed it to let the backoff kick in.